src/util/sync.js

```diff
@@ -17,12 +18,7 @@ export const sync = (thoughtIndexUpdates = {}, contextIndexUpdates = {}, { local
   // state
   // NOTE: state here is a boolean value indicating whether to sync to state
   if (state) {
-    store.dispatch({
-      type: 'thoughtIndex',
-      thoughtIndexUpdates,
-      contextIndexUpdates,
-      forceRender
-    })
+    store.dispatch(updateThoughtIndex({ thoughtIndexUpdates, contextIndexUpdates, forceRender }))
```

Comment:
Notice `if (state) { ... }`. This is important, as `state` is one of three options `state`, `local`, and `remote` that can be passed to `sync` to control where the data is persisted. For example, when a value comes back from the `remote` database (Firebase), `sync` is called with `remote: false` to avoid sending the value back up to the Firebase.

There was an assumption that `updateThoughtIndex` only affected `state`. By adding the `localStorage` logic to `updateThoughtInsdex`, unfortunately we break that invariant and cannot control the persistence targets of `sync` any longer.

Reply:
> Notice `if (state) { ... }`. This is important, as `state` is one of three options `state`, `local`, and `remote` that can be passed to `sync` to control where the data is persisted. For example, when a value comes back from the `remote` database (Firebase), `sync` is called with `remote: false` to avoid sending the value back up to the Firebase.
I'm not sure I understand what you mean. Can you please point to the diff and make it more explicit?

> There was an assumption that `updateThoughtIndex` only affected `state`. By adding the `localStorage` logic to `updateThoughtInsdex`, unfortunately we break that invariant and cannot control the persistence targets of `sync` any longer.
updateThoughtIndex is an action-creator and I believe that's where we can handle all the side-effects and API calls, including updates to the localStorage. Please let me know if you think otherwise.

Reply:
> > Notice `if (state) { ... }`. This is important, as `state` is one of three options `state`, `local`, and `remote` that can be passed to `sync` to control where the data is persisted. For example, when a value comes back from the `remote` database (Firebase), `sync` is called with `remote: false` to avoid sending the value back up to the Firebase.
> 
> I'm not sure I understand what you mean. Can you please point to the diff and make it more explicit?

Yup, I accidentally left out the first line in my original highlight!
https://github.com/cybersemics/em/pull/462/files#diff-c80007206b39ed535d2456ee5a159d1fR20-R22

The point is that you are modifying `local` inside something that is supposed to modify `state` only. Mixing these up invalidates the persistence target logic (the logic that allows `sync` to target `state`, `local`, or `remote` data stores).

> 
> > There was an assumption that `updateThoughtIndex` only affected `state`. By adding the `localStorage` logic to `updateThoughtInsdex`, unfortunately we break that invariant and cannot control the persistence targets of `sync` any longer.
> 
> updateThoughtIndex is an action-creator and I believe that's where we can handle all the side-effects and API calls, including updates to the localStorage. Please let me know if you think otherwise.

`updateThoughtIndex` already has an action-creator: `sync`. That's where the logic should go. We should move `sync` to the `action-creators` folder to make that more explicit.